### PR TITLE
Add thread pool option

### LINKED ISSUE #14025: Add thread_pool option in RSync library
# Description
This issue aims to add an thread pool limit option for the dispatch queue workers..

This limit must be configurable in the rsync instance, for the module that wants to implement it.

# DoD
- [x] Modify C interface to receive a thread pool value, use unsigned int limit. rsync.h  (EXPORTED RSYNC_HANDLE rsync_create())
- [x] Add another/overload constructor for C++ interface, to receive a thread pool value argument.
- [x] Extend RSYNC_HANDLE RSyncImplementation::create() to receive the thread pool value argument.
- [x] Modify C and C++ interface calls (in RSync::Create).
- [x] Extend RSyncContext constructor to receive the thread pool number argument.
- [x] Extend/overload and receive in the MsgDispatcher constructor a thread pool argument.

### PR DESCRIPTION
**Pull request by Dwordcito**
|Related issue|
|---|
|Closes #14025 |

## Description
This issue aims to add an thread pool limit option for the dispatch queue workers..

This limit must be configurable in the rsync instance, for the module that wants to implement it.

### CODE REVIEW

**Review by pereyra-m — COMMENTED** [review]
GJ!
Just one question

**Review by pereyra-m — APPROVED** [review]
LGTM!

**Review by pereyra-m — DISMISSED** [review]

File: src/shared_modules/utils/pipelineNodesImp.h [hunk 1]
```diff
@@ -31,7 +31,7 @@ namespace Utils
             {}
             ReadNode(Functor functor,
                      const unsigned int numberOfThreads)
-                : DispatcherType{ functor, numberOfThreads }
+                : DispatcherType{ functor, numberOfThreads, 0 }
```
Comment (HanesSciarrone) [inline src/shared_modules/utils/pipelineNodesImp.h:34]: Hi @Dwordcito 

Maybe the `UNLIMITED_QUEUE_SIZE` macro could be moved to a common file and used here as well.

File: src/shared_modules/utils/pipelineNodesImp.h [hunk 2]
```diff
@@ -63,7 +63,7 @@ namespace Utils
             {}
             ReadWriteNode(Functor functor,
                           const unsigned int numberOfThreads)
-                : DispatcherType{ std::bind(&RWNodeType::doTheWork, this, std::placeholders::_1), numberOfThreads }
+                : DispatcherType{ std::bind(&RWNodeType::doTheWork, this, std::placeholders::_1), numberOfThreads, 0 }
```
Comment (HanesSciarrone) [inline src/shared_modules/utils/pipelineNodesImp.h:66]: Idem

File: src/shared_modules/rsync/include/rsync.hpp [hunk 3]
```diff
@@ -26,30 +26,34 @@
...
 
     /**
-     * @brief Remote sync initializes the instance. 
+     * @brief Remote sync initializes the instance.
+     *
+     * @param threadPoolSize Size of the thread pool.
+     * @param maxQueueSize Maximum size of the queue, if the value is 0, it is unlimited.
      */
-    RemoteSync();
+    RemoteSync(const unsigned int threadPoolSize = std::thread::hardware_concurrency(), const size_t maxQueueSize = 0);
```
Comment (HanesSciarrone) [inline src/shared_modules/rsync/include/rsync.hpp:51]: Hi @Dwordcito

Maybe the `UNLIMITED_QUEUE_SIZE` macro could be moved to a common file and used here as well.

File: src/shared_modules/rsync/src/rsyncImplementation.h [hunk 4]
```diff
@@ -97,7 +98,8 @@ namespace RSync
 
             void releaseContext(const RSYNC_HANDLE handle);
 
-            RSYNC_HANDLE create();
+            RSYNC_HANDLE create(const unsigned int threadPoolSize = std::thread::hardware_concurrency(),
+                                const size_t maxQueueSize = 0);
```
Comment (HanesSciarrone) [inline src/shared_modules/rsync/src/rsyncImplementation.h:102]: Hi @Dwordcito

Maybe the `UNLIMITED_QUEUE_SIZE` macro could be moved to a common file and used here as well.

File: src/shared_modules/utils/msgDispatcher.h [hunk 5]
```diff
@@ -32,8 +32,14 @@ namespace Utils
         , public RawValueDecoder
     {
         public:
-            MsgDispatcher()
-                : ThreadType{std::bind(&DispatcherType::dispatch, this, std::placeholders::_1)}
+            MsgDispatcher(const unsigned int threadPoolSize = std::thread::hardware_concurrency(),
+                          const size_t maxQueueSize = 0)
```
Comment (HanesSciarrone) [inline src/shared_modules/utils/msgDispatcher.h:36]: Hi @Dwordcito

Maybe the `UNLIMITED_QUEUE_SIZE` macro could be moved to a common file and used here as well.

**Review by jnasselle — COMMENTED** [review]
GJ! I have only one comment about `std::thread::hardware_concurrency`

Documentation tells that `if the value is not well defined or not computable, returns ​0`.
In this case, no AsyncDispatcher will be created. How to improve this, in this method fails, assign one thread as a fallback?

**Comment by Dwordcito** [comment]
@jnasselle It seems to me a very good option.

**Review by jnasselle — APPROVED** [review]

**Review by HanesSciarrone — DISMISSED** [review]

**Review by jnasselle — APPROVED** [review]

**Review by jftuduri — DISMISSED** [review]

**Review by pereyra-m — APPROVED** [review]

**Review by HanesSciarrone — APPROVED** [review]

**Review by jnasselle — APPROVED** [review]

**Review by pereyra-m — APPROVED** [review]

**Review by MiguelazoDS — APPROVED** [review]

**Pull request merged**